### ISSUE TITLE
NPE at the start of firing phase

### ISSUE THREAD

**Issue by Question2** [MegaMek/megamek#644]
Couldnt hit done because there was a NPE : 

```
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
	at megamek.common.Game.getNextEntityNum(Game.java:1843)
	at megamek.common.Game.getNextEntity(Game.java:1813)
	at megamek.client.ui.swing.FiringDisplay.endMyTurn(FiringDisplay.java:919)
	at megamek.client.ui.swing.FiringDisplay.ready(FiringDisplay.java:1292)
	at megamek.client.ui.swing.AbstractPhaseDisplay$1.actionPerformed(AbstractPhaseDisplay.java:118)
	at javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
	at javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.setPressed(Unknown Source)
	at javax.swing.plaf.basic.BasicButtonListener.focusLost(Unknown Source)
	at java.awt.Component.processFocusEvent(Unknown Source)
	at java.awt.Component.processEvent(Unknown Source)
	at java.awt.Container.processEvent(Unknown Source)
	at java.awt.Component.dispatchEventImpl(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.KeyboardFocusManager.redispatchEvent(Unknown Source)
	at java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(Unknown Source)
	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(Unknown Source)
	at java.awt.Component.dispatchEventImpl(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.SentEvent.dispatch(Unknown Source)
	at java.awt.DefaultKeyboardFocusManager$DefaultKeyboardFocusManagerSentEvent.dispatch(Unknown Source)
	at java.awt.DefaultKeyboardFocusManager.sendMessage(Unknown Source)
	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(Unknown Source)
	at java.awt.Component.dispatchEventImpl(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Window.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)
```

[#battletech.zip](https://github.com/MegaMek/megamek/files/1291585/battletech.zip)

**Comment by NickAragua** [comment]
I think I've encountered this, so I'll work on fixing it.

**Comment by arlith** [comment]
There's a lot that's going on in this ticket.  The first and biggest issue is that simultaneous fire is turned on with Princess.  Princess does support simultaneous phases and I think this is what ultimately leads to all of the other problems (there's a SF bug related to this issue).

For the reported issue, it would appear that ultimately the turn index got set to an invalid value (probably the end of the turn vector), which caused a null turn to get returned for `Game.getNextEntityNum`.  I'm not 100% certain on what situation would lead to this, as it seems odd.  It's easy enough to fix this with a null-guard, but I'm not certain that's the right answer here.  It'd be better to understand the issue, but I'm positive this is due to a race condition so reproducing it is difficult.

There's another issue present in the logfile, represented by this block:

>error: server got invalid attack packet from connection 1, Entity: Phantom D
error: server got invalid attack packet from connection 1, Entity: Phantom D #2
error: server got invalid attack packet from connection 1, Entity: Ryoken (Stormcrow) B
error: server got invalid attack packet from connection 0, Entity: Ryoken (Stormcrow) A 2 #2
error: server got invalid attack packet from connection 1, Entity: Ryoken (Stormcrow) B
error: server got invalid attack packet from connection 1, Entity: Puma (Adder) C
error: server got invalid attack packet from connection 1, Entity: Puma (Adder) C
error: server got invalid attack packet from connection 1, Entity: Puma (Adder) H
error: server got invalid attack packet from connection 1, Entity: Puma (Adder) H
error: server got invalid attack packet from connection 1, Entity: Puma (Adder) H
error: server got invalid attack packet from connection 0, Entity: Ryoken (Stormcrow) A 2
error: server got invalid attack packet from connection 1, Entity: Koshi (Mist Lynx) D

The connection 1 invalid attack packets aren't a big deal, this is a part of the issue with Princess not working with simultaneous phases.  The surprising thing is the connection 0 invalid attack packets.

This comes from Server line 13312, and happens if the Server receives an attack action from a client that isn't valid for the current turn.  In simultaneous phases, the turn is picked as the first turn for  the player who sends the packet.  In simultaneous phases this message almost always means that the Server got an attack packet for a unit that has already attacked.  So what this is saying is that, somehow, a players client is sending an attack from the same unit twice.

I think I've figured out how this could happen.  Basically, if the Bot client sends a done action for one unit, just before the player sends a done action for their unit, the players client will receive the result of bot clients action before it receives the result of it's own action.  This may very well lead to the client selecting the same unit it just sent an attack for, allowing the client to send another action for that unit, resulting in the invalid attack packet.  This would be more likely if the player is just rapidly clicking done (or hitting ctrl-enter) to cycle through all units for a "do nothing" round.

Ultimately, I think that the client may need to update `Entity` state.  I think the client can indicate to itself that a unit is done by marking it locally done after sending the action to the server.  I think we've generally avoided doing this, because the Server ends up overwriting this state once it transmits the results.  I'm not really sure this solves the problem though.  Marking the Entity as done will prevent the client from reselecting it, but it doesn't update the `Game` instances turn information, and I'm reluctant to update that because it's complicated to get right.

It may be the case that we need a better mechanism for indicating whose turn it is during sim play.  Ultimately, the problem is that, every time a turn ends, it fires a `GameTurnChangeEvent`, which signals the phase displays.  If the new turn is your turn, and you're not taking a turn, start your turn, basically.  This mechanism works fine in sequential play, but it wasn't designed for simultaneous play and just got repurposed.

In the case I described above, the issue is that, the phase display doesn't expect to receive information about a turn starting before the server has finished processing the turn it just sent.

I think the most straightforward way to deal with the issue is to add information about who took the turn that's triggering the `GameTurnChangeEvent`.  I think this solves the aforementioned problem and also solves the Princess sim phase problem as well (ultimately the issue there is that everytime a `gameTurnChagneEvent` happens, Princess fires off a new thread to calculate a turn for a unit.  Since this triggers on everyone's turn, Princess ultimately spawns way more threads than it needs to, and this can lead to it trying to move the same unit multiple times).
